Update Media Card to check for Supported Features

zsarnett opened the pull request:
## Type of change
- [x] Bugfix (non-breaking change which fixes an issue)
- [x] Code quality improvements to existing code or addition of tests

- This PR fixes or closes issue: fixes #4844

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

zsarnett committed:
Update Media Card to check for supported features

bramkragten commented:
Still some big differences, mainly:
- Play button state doesn't get disabled
- Unavailable state -> Play button + image color
- Size of album covers
- The progress bar doesn't disappear when stopped playing
- Animation on height change is missing
- Missing now playing info
- Long media title isn't cut off (ellipsis)

Before:
![image](https://user-images.githubusercontent.com/5662298/74337100-f6901d00-4d9f-11ea-8abc-fd68d6086f8e.png)
![media-before](https://user-images.githubusercontent.com/5662298/74338318-407a0280-4da2-11ea-9e8b-d6619e654cfe.gif)

After:
![image](https://user-images.githubusercontent.com/5662298/74336881-7cf82f00-4d9f-11ea-856e-04cfa8f3c1b1.png)
![media-after](https://user-images.githubusercontent.com/5662298/74338469-88008e80-4da2-11ea-8a0b-3e385433b3c9.gif)

zsarnett commented:
> * Play button state doesn't get disabled

Fixed. Disabled when the state is `off`

> * Unavailable state -> Play button + image color

Fixed. Play button is no longer rendered and the image is darkened when the state is `unavailable`

> * Size of album covers

This is something I changed to be consistent with the card size. I have made it so all the albums are the same size if the card is playing

> * The progress bar doesn't disappear when stopped playing

I disagree with the progress bar being removed if the state is `paused`. I would still expect the player to know my place and continue when I hit play.

> * Animation on height change is missing

Fixed. Height is animated when the image is changed

> * Missing now playing info

Fixed. This was changed the be `Secondary_Info: Primary_Info` when the card was converted. I have reverted this change.

> * Long media title isn't cut off (ellipsis)

Fixed. Media Title is not set to show ellipsis when longer than one row of text

zsarnett committed:
Review Updates

zsarnett committed:
Disabled Style Updates

zsarnett committed:
Support for requesting the image from the server

zsarnett commented:
@balloob The demo (Arsaboo) has the image pulling from api folder. The new system tries to find that and convert to base64 but for that image, in particular, it isn't found. Resulting in a blue background

zsarnett committed:
Stupid mistake

balloob commented:
I don't understand, we shouldn't try to fetch the image unless there is one? Also why are we converting it to base64? It's already a url. If it needs auth we should use a signed url.

zsarnett commented:
The reason the image shows up blank before the recent change was that an entity image was supplied "api/.../image" URL. But that URL was not found and no image was returned. The previous way doesn't check to see if that image exists or not. We are only converting to base64 if the image is a local image.

zsarnett commented:
#4489 Also Adding this here as something that may need to be checked for. Will use this PR to update

bramkragten commented on src/panels/lovelace/cards/hui-media-control-card.ts:
```diff
@@ -68,28 +78,41 @@ export class HuiMediaControlCard extends LitElement implements LovelaceCard {
+          ></div>
+          <div
+            class="${classMap({
+              caption: true,
```
Don't add static values to a `classMap`:
```
class="caption ${classMap({...})"
```

bramkragten commented on src/panels/lovelace/cards/hui-media-control-card.ts:
```diff
@@ -99,35 +122,59 @@ export class HuiMediaControlCard extends LitElement implements LovelaceCard {
+                `}
+            ${(stateObj.state !== "playing" &&
+              !supportsFeature(stateObj, SUPPORTS_PLAY)) ||
+            stateObj.state === "unavailable"
```
```suggestion
            stateObj.state === "unavailable" || (stateObj.state == "playing" && !supportsFeature(stateObj, SUPPORT_PAUSE) &&  !supportsFeature(stateObj, SUPPORT_STOP))
```

bramkragten replied:
Use the const for `unavailable` -> https://github.com/home-assistant/home-assistant-polymer/blob/dev/src/data/entity.ts#L1

bramkragten commented on src/panels/lovelace/cards/hui-media-control-card.ts:
```diff
@@ -99,35 +122,59 @@ export class HuiMediaControlCard extends LitElement implements LovelaceCard {
           </div>
           <paper-icon-button
-            icon="hass:dots-vertical"
+            .icon=${"hass:dots-vertical"}
```
Only make it a property when it dynamic

bramkragten commented on src/panels/lovelace/cards/hui-media-control-card.ts:
```diff
@@ -158,32 +205,53 @@ export class HuiMediaControlCard extends LitElement implements LovelaceCard {
+      return;
+    }
+
+    fetchMediaPlayerThumbnailWithCache(this.hass, this._config.entity).then(
```
@balloob didn't we move the thumbnails away from the ws?

zsarnett replied:
What is the status on how we should fetch the image? Should it be reverted to using the image URL or is this the best method. This is the method it was using before the conversion.

zsarnett committed:
Review updates

bramkragten commented on src/panels/lovelace/cards/hui-media-control-card.ts:
```diff
@@ -129,7 +130,7 @@ export class HuiMediaControlCard extends LitElement implements LovelaceCard {
             : html`
                 <paper-icon-button
-                  .icon=${"hass:power"}
+                  icon=${"hass:power"}
```
```suggestion
                  icon="hass:power"
```

bramkragten commented on src/panels/lovelace/cards/hui-media-control-card.ts:
```diff
@@ -139,15 +140,18 @@ export class HuiMediaControlCard extends LitElement implements LovelaceCard {
               : html`
                   <paper-icon-button
-                    .icon=${"hass:skip-previous"}
+                    icon=${"hass:skip-previous"}
```
```suggestion
                    icon="hass:skip-previous"
```

bramkragten commented on src/panels/lovelace/cards/hui-media-control-card.ts:
```diff
@@ -166,15 +170,15 @@ export class HuiMediaControlCard extends LitElement implements LovelaceCard {
               : html`
                   <paper-icon-button
-                    .icon=${"hass:skip-next"}
+                    icon=${"hass:skip-next"}
```
```suggestion
                    icon="hass:skip-next"
```

bramkragten commented on src/panels/lovelace/cards/hui-media-control-card.ts:
```diff
@@ -166,15 +170,15 @@ export class HuiMediaControlCard extends LitElement implements LovelaceCard {
           </div>
           <paper-icon-button
-            .icon=${"hass:dots-vertical"}
+            icon=${"hass:dots-vertical"}
```
```suggestion
            icon="hass:dots-vertical"
```

bramkragten commented on src/panels/lovelace/cards/hui-media-control-card.ts:
```diff
@@ -166,15 +170,15 @@ export class HuiMediaControlCard extends LitElement implements LovelaceCard {
           <paper-icon-button
-            .icon=${"hass:dots-vertical"}
+            icon=${"hass:dots-vertical"}
             @click="${this._handleMoreInfo}"
```
Also we should drop the `""` on dynamic values.
```suggestion
            @click=${this._handleMoreInfo}
```

zsarnett committed:
Review updates again because I am a fool

zsarnett committed:
Update for playlist information if music isnt present

zsarnett commented:
Fixes: #4503

zsarnett committed:
When no image occurs mark the image as undefined

zsarnett committed:
rookie mistake

zsarnett committed:
Typing functions
